### PR TITLE
Avoid unnecessary switch to `MainConnection` while setting `read-only` mode

### PR DESCRIPTION
**Pull request by wyrzyk**
As per JavaDoc `read-only` mode is just a hint to the driver to enable
optimisations. In the case of `DualConnection` switching to the main just in case
is anti-optimisation.

I found out we set it to `false` just in case in `DatabaseAccessorImpl#runInTransaction`.
There might be more places like this, that prevent us from utilising replica more.

### CODE REVIEW

File: src/test/java/com/atlassian/db/replica/api/TestDualConnection.java [hunk 1]
```diff
@@ -783,25 +783,39 @@ public void shouldStopReadOnlyMode() throws SQLException {
...
         assertThat(connectionProvider.getProvidedConnectionTypes())
-            .containsOnly(MAIN);
+            .containsOnly(REPLICA);
         verify(connectionProvider.singleProvidedConnection()).setReadOnly(false);
         assertThat(connection.isReadOnly()).isFalse();
     }
 
     @Test
-    public void shouldNotSwitchBackToReplica() throws SQLException {
+    public void shouldUtiliseReplicaEvenAfterDisablingReadOnly() throws SQLException {
```
Comment (dagguh) [inline src/test/java/com/atlassian/db/replica/api/TestDualConnection.java:792]: Sweet

File: src/test/java/com/atlassian/db/replica/api/TestDualConnection.java [hunk 2]
```diff
@@ -783,25 +783,39 @@ public void shouldStopReadOnlyMode() throws SQLException {
...
         connection.setReadOnly(false);
-        connection.setReadOnly(true);
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        assertThat(connectionProvider.getProvidedConnectionTypes())
+            .containsOnly(REPLICA);
+    }
+
+    @Test
+    public void shouldNotSwitchBackToReplica() throws SQLException {
```
Comment (dagguh) [inline src/test/java/com/atlassian/db/replica/api/TestDualConnection.java:807]: This test case seems unrelated to the PR. Correct?

**Review by dagguh — APPROVED** [review]

File: src/test/java/com/atlassian/db/replica/api/TestDualConnection.java (hunk 2 above)
Comment (wyrzyk) [reply src/test/java/com/atlassian/db/replica/api/TestDualConnection.java:807]: Yes. The test case is unrelated. The diff doesn't show it, but   `shouldNotSwitchBackToReplica` was reusing `setReadOnly` to switch between the main database and replica. It will not work after the fix, so I changed it to use `executeUpdate` and `executeQuery`.
